pip/pip-385.md

```diff
@@ -0,0 +1,398 @@
+
+# Detailed Design
+
+## High-level Implementation Details
+
+### Broker Changes
+
+* For calls arising from `PublishRateLimiterImpl` class, add logic in `ServerCnxThrottleTracker.java` to send the
+  command to client and wait for response for the max configured duration before calling `changeAutoRead`. It checks for
+  feature availability first.
+* For calls arising from `ServerCnxThrottleTracker::changeThrottlingFlag`, we send the command async (if feature
+  supported) without worrying about response and then call `changeAutoRead`.
+* capture `AbstractTopic::getTotalPublishRateLimitCounter` per publish rate limit counter and add relevant attribute in
+  the rate limit metric
+
+### Determining the throttling duration for clients
+
+Broadly, there are 2 different categories for rate limiting - (a) quota exceeding at a topic/resource group level OR (b)
+connection level breaches leading to broker's resource crunch.
+
+In case (a), as per the current token bucket based quota tracking introduced in PIP-322, we (lazily) resume channel
+reads on the next second mark. i.e. the throttling is only applied at max for 1 second.
+In case (b), there is no clear picture on when the channel read is resumed as we wait till we have processed half of the
+pending messages, or cleared half of the memory threshold.
+
+Thus, for case (a), we will be asking the client to pause further message sending for the remainder of the second and in
+case (b), we will send `0` as the channel read is anyway paused.
+
+### Java Client Changes
+
+* We maintain a boolean and an enum field at a producer level to capture if the producer is being throttled or not and
+  the reason for current throttling.
+* We also capture the timestamp at which we began throttling, and the duration for which we need to continue throttling
+  from that timestamp.
+* `ClientCnx` to handle incoming `ThrottleProducer` command and set the boolean and timestamp for the relevant producer
+  and send back a the `ThrottleProducerReceipt`.
+    * We mark a counter metric with the reason and the value being the throttle duration.
+* Most of the blocking logic lies in `ProducerImpl::processOpSendMsg`. More details in next section. 
+* Producer disables the throttling lazily, whenever next time there is a check for "being throttled" and required time
+  has passed.
+* In the timeout processor `run` method, we add logic to send a rate limit exception if the producer is being throttled
+  instead of TimeoutException. Explained in further sections.
+* For partitioned producer - `isThrottled` returns true if any of the partitions is being throttled.
+* For partitioned producer - In case of default round robing partition selection approach, we skip the partition which
+  is being throttled.
+
+### Blocking messages to be sent during throttling
+
+Most of the logic to actually block a message until throttling is lifted resides inside `ProducerImpl::processOpSendMsg`
```

Comment:
In the current client, this blocking approach is problematic since it also applies to the Pulsar Java client async API. Blocking threads in asynchronous code is something to avoid and there's currently not a great way to do that in the Pulsar Java client. 
In the Pulsar Reactive Java client, it required a relatively complex solution to implement non-blocking backpressure on top of the Pulsar Java client async API. It's handled by keeping a limited amount of messages in flight. The way to implement that isn't trivial for Reactive Streams / Project Reactor (one part of implementation is in [InflightLimiter class](https://github.com/apache/pulsar-client-reactive/blob/main/pulsar-client-reactive-api/src/main/java/org/apache/pulsar/reactive/client/internal/api/InflightLimiter.java)). 
It would be great if the Pulsar Java client async API had a non-blocking way for producer flow control. 
This is a broad problem that needs a separate PIP.